cases/src/org/commcare/cases/instance/CaseDataInstance.java

```diff
@@ -0,0 +1,84 @@
+package org.commcare.cases.instance;
+
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.model.instance.utils.FormLoadingUtils;
+import org.javarosa.core.services.Logger;
+import org.javarosa.xml.util.InvalidStructureException;
+
+import java.io.IOException;
+
+/**
+ * An external data instance that respects CaseDB template specifications.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class CaseDataInstance extends ExternalDataInstance {
+    private static TreeElement caseDbSpecTemplate;
+    private static final String CASEDB_WILD_CARD = "CASEDB_WILD_CARD";
+
+    static {
+        try {
+            caseDbSpecTemplate =
+                    FormLoadingUtils.xmlToTreeElement("/casedb_instance_structure.xml");
+        } catch (InvalidStructureException e) {
+            caseDbSpecTemplate = null;
+        } catch (IOException e) {
+            caseDbSpecTemplate = null;
+        }
+    }
+
+    public CaseDataInstance() {
+        // For serialization
+    }
+
+    private CaseDataInstance(String reference, String instanceid) {
+        super(reference, instanceid);
+    }
+
+    public CaseDataInstance buildExternalDataInstance(String reference, String instanceId) {
+        return new CaseDataInstance(reference, instanceId);
+    }
+
+    /**
+     * Does the reference follow the statically defined CaseDB spec?
+     */
+    public boolean hasTemplatePath(TreeReference ref) {
+        if (caseDbSpecTemplate != null) {
+            return followsTemplateSpec(ref, caseDbSpecTemplate, 0);
+        } else {
+            // failed to load casedb spec template, default to super implementation
+            Logger.log("CaseDb Warning",
+                    "Using default hasTemplatePath implementation: reference resolution will break!");
+            return super.hasTemplatePath(ref);
```

Comment:
Honestly I'm thinking whether it would make more sense to literally inline the template text as static content here, although that would make the permgen size of this file really large. 

It makes me really uncomfortable that the code path for loading this is also static, as it introduces _huge_ unpredictable hits across the codebase for when a complex parse happens (And will/can mean it happens at weird and bad times), can we make the static init here happen lazily instead? (Is fine for the template parse itself to be static, I think), as long as it is stateless and threadsafe)